### ISSUE TITLE
Email logins cause IP lockout if "Instantly Lockout Invalid Usernames" feature is active

### ISSUE THREAD

**Issue by chesio** [Arsenal21/all-in-one-wordpress-security#66]
WordPress 4.5 introduced an option to log with email. This is basically done via an additional `authenticate` filter, see the [changeset](https://core.trac.wordpress.org/changeset/36617/trunk/src/wp-includes/default-filters.php).

AIOWPSF [replaces](https://github.com/Arsenal21/all-in-one-wordpress-security/blob/master/all-in-one-wp-security/classes/wp-security-user-login.php#L13-L14) the default `authenticate` filter to perform its login-related functions, but fails to address the new filter that authenticates username as email. Therefore, if [someone tries to log in with email](https://wordpress.org/support/topic/email-logins-cause-ip-lockouts-with-specific-aio-setting/) and "Instantly Lockout Invalid Usernames" feature is active, the new filter logs him in, but still a new record is added to login lockdown table, because default re-implementation only [checks](https://github.com/Arsenal21/all-in-one-wordpress-security/blob/master/all-in-one-wp-security/classes/wp-security-user-login.php#L105) `$username` against username field, not email.

To be honest, I don't like the current solution at all. I don't think the default `authenticate` filter should be replaced, it makes the implementation error prone with respect to future changes (this issue is a nice example). In my opinion, the plugin should hook before and after default authentication filters to perform its checks. If you see it the same way and are fine with a more thorough refactorization of related code, I'd be happy to work on a fix - I have some time in coming days. If you don't, then I'll probably refactor the code anyway, but I'm not going to waste my time with a pull request.

Btw. several other methods in `AIOWPSecurity_User_Login` class likely need to be updated as well.

**Comment by Arsenal21** [comment]
The new version of the plugin has a fix for this.

**Comment by chesio** [comment]
No, as far as I can see (and from what I have tested), the new version of the plugin only makes it impossible to log in with email (and immediately blocks an IP, if someone tries to log in with email and "Instantly Lockout Invalid Usernames" feature is active).

**Comment by chesio** [comment]
Well, I'm not the only one who thinks the fix in 4.2.2 actually [made the things worse](https://wordpress.org/support/topic/email-logins-cause-ip-lockouts-with-specific-aio-setting/#post-8448706).

Anyway, I have a patch ready, but it's a [big change](https://github.com/Arsenal21/all-in-one-wordpress-security/compare/master...chesio:authentication-fixes), so I'm not going to issue a PR. If you're interested, take your time to test the [patched branch](https://github.com/chesio/all-in-one-wordpress-security/tree/authentication-fixes) and let me know. I already tested the changes in my environment, but I'm not using "rename login page" and "manual registration approval" features, so I might missed something.

Despite being a large change, the patch retains existing behaviour with two notable exceptions:

1. If user IP is locked, the error message and "Request Unlock" button (if enabled) are displayed via call to [wp_die()](https://developer.wordpress.org/reference/functions/wp_die/) with status 503 instead within the login form. This is done in order to stop further authentication hooks from processing, but I also consider returning 503 error more apropriate in such case (I got inspired by Wordfence here to be honest). Also, it makes no sense to display login form again in case of lockout.
1. I found out that login captcha can be easily bypassed - one (bot) has just to omit all captcha data and also "wp-submit" from _POSTed data. Login action is triggered even without `$_POST['wp-submit']`, but plugin only returns captcha error if this variable is set. This is fixed in my patch too.

In case of questions, I'm at your disposal.
